### ISSUE TITLE
accuracy regression for resnet50 in test_train_imagenet.py

### ISSUE THREAD

**Issue by zcain117** [pytorch/xla#1025]
I've tried 2 ways, both using the real imagenet dataset. I have not submitted https://github.com/pytorch/xla/pull/1012 with the resnet50 learning rate scheduler:
1. Brand new compute VM, using pytorch-nightly conda env
2. Brand new compute VM, pull from master on Github and run build_torch_wheels.sh

Both versions reach ~12% accuracy around epoch 2 or 3 and can't make it any higher no matter how many more epochs they train.

I have a different compute VM made on August 28 that uses pytorch-nightly conda env and on that compute VM I am able to reach 60% accuracy with no changes and 76% accuracy when I implement a learning rate schedule.

**Comment by zcain117** [comment]
Ran a diff of /usr/share/torch-xla-nightly/pytorch/vision/torchvision/models/resnet.py on the old VM that works and the new VM that does not and found no difference

**Comment by zcain117** [comment]
confirmed that test_train_cifar.py passes using the in-line version of resnet18 but fails to reach 80% accuracy using the torchvision version of resnet18

**Comment by taylanbil** [comment]
did we rule out the lr_scheduler PR? IIRC, high accuracy run is w/o that pr, and the low accuracy is with, is that correct?

**Comment by zcain117** [comment]
> did we rule out the lr_scheduler PR? IIRC, high accuracy run is w/o that pr, and the low accuracy is with, is that correct?

No, I've been using master version of pytorch/xla and haven't checked in that PR

**Comment by dlibenzi** [comment]
So it seems like a torchvision change?

**Comment by zcain117** [comment]
> So it seems like a torchvision change?

there was no difference in /usr/share/torch-xla-nightly/pytorch/vision/torchvision/models/resnet.py between the older, higher accuracy compute VM and the new compute VMs with bad accuracy

**Comment by zcain117** [comment]
I've run a lot of experiments and here are the 3 that I think are most enlightening:

1. Copied the image from the high-performing compute VM originally created August 28. Built new torch and torch_xla wheels and torchivision using update_nightly_torch_wheels.sh. Ran against a new TPU VM. This made it to ~45% accuracy on epoch 7, i.e. aligned with the trajectory that gets us to 76% final accuracy.
2. New compute VM built using build_torch_wheels.sh and a new TPU VM. Whether I run on master branch or on the branch that pulls #1012, accuracy can't make it past ~12-15%
3. New compute VM using pytorch-nightly conda env and a new TPU VM made it to ~43% by epoch 11. Looks promising

From this, it seems building from head using build_torch_wheels.sh results in test_train_imagenet.py not working for some reason.

I'll keep investigating

**Comment by ailzhang** [comment]
@zcain117 I'm a little confused. I thought we use `build_torch_wheels.sh` to build nightly conda env, no? 
fyi https://github.com/pytorch/xla/blob/master/scripts/build_torch_wheels.sh#L98 fixes it to torchvision 0.3.0, maybe check whether nightly has master or 0.3.0 to see if that makes a difference?

**Comment by dlibenzi** [comment]
Great! :tada: 
But we still have a mystery to solve.
Zach is getting bad accuracy. Could it be the TPU VMs he is using?

**Comment by jysohn23** [comment]
I think its in the optimizer PR.

**Comment by dlibenzi** [comment]
> I think its in the optimizer PR.

But Zach said he tested vanilla `test_train_imagenet.py` ...

**Comment by jysohn23** [comment]
Yes and I think there was no problem with vanilla test_train_imagenet.py.
Vanilla just doesnt get that high of accuracy w/o LR scheduling by epoch
20. With the rampup and stepping we are now at 73%+ top 1 at 77th epoch.
Will update later today at 90th epoch.

**Comment by zcain117** [comment]
I think the potential difference is that unmodified `test_train_imagenet.py` no longer hits 50-60% accuracy, or at least not on the schedule it did before (high 40's in by epoch 20 and then slowly creeps up to 50-60%).

From a recent unmodified run:
> [xla:1] Accuracy=13.80%
> [xla:4] Accuracy=15.82%
> [xla:7] Accuracy=13.97%
> [xla:5] Accuracy=16.61%
> [xla:6] Accuracy=12.18%
> [xla:8] Accuracy=10.70%
> [xla:2] Accuracy=13.90%
> [xla:3] Accuracy=19.30%
> Epoch: 33, Mean Accuracy: 14.53%

However, with a learning rate scheduler (in this case, we were using [this hacky version](https://github.com/pytorch/xla/tree/lr-hack), there seems to be no issues with accuracy.

I don't think the regression of unmodified resnet50 is worth spending more time on - maybe the 50-60% I was getting before was just lucky. The convergence is a lot more reliable with a learning rate scheduler

**Comment by zcain117** [comment]
Previously, I was pulling in the [PR for the LR scheduling](https://github.com/pytorch/xla/pull/1012), running `build_torch_wheels.sh`, and then running `test_train_imagenet.py` but I was getting low accuracy.

With Daniel's finding, I think this narrows down our current problem to 1 of 2 things:
1. There is a mistake in that PR
2. Something strange happens with `build_torch_wheels.sh`

I'm going to try again now using 2 new compute VMs, once where I pull in the PR and build from scratch with build_torch_wheels.sh and once where I patch the changes in directly to the conda environment source code

**Comment by jysohn23** [comment]
OK great! We're good and hit 76.3% accuracy at 90th epoch.

<img width="510" alt="Screen Shot 2019-09-13 at 11 30 57 AM" src="https://user-images.githubusercontent.com/19496130/64886199-7318a700-d61a-11e9-99ce-cfd56b97fffd.png">

Full logs: 
[float32_batch128_lr0.5_divideBy5Every20Epochs_warmup1EpochLinear (2).txt](https://github.com/pytorch/xla/files/3611377/float32_batch128_lr0.5_divideBy5Every20Epochs_warmup1EpochLinear.2.txt)

**Comment by zcain117** [comment]
There were 5 tests I wanted to run:
1. Run using latest pytorch-nightly with [https://github.com/pytorch/xla/pull/1012](new learning rate scheduler) patched in
2. Build from head using build_torch_wheels.sh after pulling [https://github.com/pytorch/xla/pull/1012](new learning rate scheduler)
3. Run with no learning rate scheduler using latest pytorch-nightly
4. Build from head and run with no learning rate scheduler
5. Run with no learning rate scheduler using August 28 pytorch-nightly

Results:
1. 76.2% accuracy [latest_pytorch_nightly_including_new_scheduler.txt](https://github.com/pytorch/xla/files/3613013/latest_pytorch_nightly_including_new_scheduler.txt)
2. Failed: grabbed source during a torch breakage caused by [this](https://github.com/pytorch/pytorch/pull/24352/files). Will try again since they [reverted that change](https://github.com/pytorch/pytorch/commit/a3f0d988d9cae1421aa24e215d402f708ee15aa7)
3. 15.51% accuracy
[float32_batch128_lr0.5_divideBy5Every20Epochs_warmup1EpochLinear_builtFromHead.txt](https://github.com/pytorch/xla/files/3613015/float32_batch128_lr0.5_divideBy5Every20Epochs_warmup1EpochLinear_builtFromHead.txt)
4. 15.51% accuracy
[float32_batch128_lr0.5_divideBy5Every20Epochs_warmup1EpochLinear_unmodifiedConda.txt](https://github.com/pytorch/xla/files/3613018/float32_batch128_lr0.5_divideBy5Every20Epochs_warmup1EpochLinear_unmodifiedConda.txt)
5. TODO

It seems we are good to go when using a learning rate scheduler but lower accuracy than before if not using a learning rate scheduler. I'll run test 5 to see exactly what accuracy we used to get, but from my memory it was ~40-50%.

I'll also repeat test 2 with latest source to make sure the new scheduler PR is safe to submit

**Comment by zcain117** [comment]
I was able to get the no-scheduler version up to ~45% accuracy by playing with learning rate and it looks the same whether on Sept 14 pytorch-nightly or Aug 28, so I think there is no regression.

My only remaining question is whether we reliably get the same results with build from head vs. from the installed conda version, but I'll follow up on that outside of this issue and make a new one if I see a difference
